ui/src/vayla-design-lib/closeable-modal/closeable-modal.tsx

```diff
@@ -1,59 +1,288 @@
+        position: modalPosition,
+        size: modalSize,
+    };
+
+    const overflow = calculateOverflow(viewportSize, modalRect);
+
+    return {
+        rect: modalRect,
+        overflow,
+    };
+}
+
+function findOptimalModalPosition(
+    viewportSize: Dimensions,
+    anchorElementRect: Rect,
+    anchorElementAnchor: Anchor,
+    modalSize: Dimensions,
+    modalAnchor: Anchor,
+    margin: number,
+    overflowSolvingMethod: OverflowSolvingMethod,
+) {
+    const { rect: defaultRect, overflow: defaultOverflow } = calculateModalRectAndOverflow(
+        viewportSize,
+        anchorElementRect,
+        anchorElementAnchor,
+        modalSize,
+        modalAnchor,
+        margin,
+    );
+
+    if (
+        overflowSolvingMethod === 'REPOSITION' &&
+        (defaultOverflow.width > 0 || defaultOverflow.height > 0)
+    ) {
+        const { rect: repositionedRect, overflow: repositionedOverflow } =
+            calculateModalRectAndOverflow(
+                viewportSize,
+                anchorElementRect,
+                getOppositeAnchorByOverflow(anchorElementAnchor, defaultOverflow),
+                modalSize,
+                getOppositeAnchorByOverflow(modalAnchor, defaultOverflow),
+                margin,
+            );
+
+        if (
+            repositionedOverflow.width < defaultOverflow.width ||
+            repositionedOverflow.height < defaultOverflow.height
+        ) {
+            return repositionedRect;
+        } else {
+            return defaultRect;
+        }
+    }
+
+    return defaultRect;
+}
+
+function getDefaultAnchors(
+    modalPosition: ModalPosition,
+    openTowards: OpenTowards,
+): [Anchor, Anchor] {
+    const horizontalEdgeToAlign = openTowards === 'RIGHT' ? 'LEFT' : 'RIGHT';
+    if (modalPosition === 'BELOW') {
+        return [
+            {
+                horizontal: horizontalEdgeToAlign,
+                vertical: 'BOTTOM',
+            },
+            {
+                horizontal: horizontalEdgeToAlign,
+                vertical: 'TOP',
+            },
+        ];
+    } else if (modalPosition === 'ABOVE') {
+        return [
+            {
+                horizontal: horizontalEdgeToAlign,
+                vertical: 'TOP',
+            },
+            {
+                horizontal: horizontalEdgeToAlign,
+                vertical: 'BOTTOM',
+            },
+        ];
+    } else if (modalPosition === 'LEFT') {
+        return [
+            {
+                horizontal: 'LEFT',
+                vertical: 'TOP',
+            },
+            {
+                horizontal: 'RIGHT',
+                vertical: 'TOP',
+            },
+        ];
+    } else if (modalPosition === 'RIGHT') {
+        return [
+            {
+                horizontal: 'RIGHT',
+                vertical: 'TOP',
+            },
+            {
+                horizontal: 'LEFT',
+                vertical: 'TOP',
+            },
+        ];
+    } else throw new Error('Unhandled modal position');
```

Comment:
Näin suoralta `throw`:aamiselta saattaisi välttyä esim. sillä muualla käytetyllä `switch`-`never`-rakenteella. Ainakin `map-view`:issä on karttatasojen luonnissa sellainen jos haluaa katsoa mallia.